streaming/base/dataloader.py

```diff
@@ -73,6 +73,13 @@ def state_dict(self) -> Optional[Dict[str, Any]]:
         if isinstance(self.dataset, StreamingDataset):
             world = World.detect()
             num_samples = self.num_samples_yielded * world.num_ranks
+            if self.dataset.replication is not None:
+                # Check if we are using `replication`. If we are, then we need to adjust the
+                # `num_samples_yielded` to reflect the fact that sample ids are shared across
+                # `replication` consecutive devices. For example, if `replication` is 2, then the
+                # number of samples seen is half the number of samples yielded, since every pair
+                # of devices shares sample ids. So the index into the sample partition is halved.
+                num_samples = num_samples // self.dataset.replication
```

Comment:
Why put this logic here and not in the dataset.py? Not everyone uses `StreamingDataloader`.

Reply:
So previously this logic was in `dataset.py`, but if someone is trying to get deterministic resumption, they should be using `StreamingDataLoader` anyways. The other path for people to get deterministic resumption is through Composer, which already takes care of this issue.

Reply:
`dataset.py` can't tell if the number of samples passed in is counting unique samples or not, so it's safer not to include it in the dataset.